### PR TITLE
Endpoint to create team

### PR DESCRIPTION
**Pull request by JohnFarrellDev**
## Link to JIRA ticket

[Backend API - Create Team](https://hackney.atlassian.net/jira/software/projects/SCS/boards/51?selectedIssue=SCS-797)

### *What is the problem we're trying to solve*

Currently we can't create teams via our API.
We are going to need this functionality anyway but also Creating a worker requires a team.
To debug locally we want to be able to make a team via a REST API instead of manually inserting into a database.

Basically this takes care of some future required work anyway and moves us a step closer to proper local development/debugging.

### *What changes have we introduced*

Create a new controller for Teams.

- Created a post endpoint which expects a team name, team context with validation.
- Added to the team usecase an execute post function. (todo is not allowing us to create a team when the team name is already taken, this will be done in following piece of work when we make a get team by team name API)
- Added database gateway function for saving the team to the database
- Updated entity factory to also return team context
- Created response factory method for returning a team response
- Tested all new/edited functionality

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### *Follow up actions after merging PR*

Are there any next steps that need to addressed after merging this PR? Add them here.

### CODE REVIEW

File: SocialCareCaseViewerApi/V1/Boundary/Requests/CreateTeamRequest.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    public class CreateTeamRequest
+    {
+        [JsonPropertyName("name")]
+        public string Name { get; set; } = null!;
+
+        [JsonPropertyName("context")]
+        public string Context { get; set; } = null!;
+    }
+
+    public class CreateTeamRequestValidator : AbstractValidator<CreateTeamRequest>
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi/V1/Boundary/Requests/CreateTeamRequest.cs:17]: What is `AbstractValidator`? Is that from `FluentValidation`?

File: SocialCareCaseViewerApi/V1/UseCase/TeamsUseCase.cs [hunk 2]
```diff
@@ -21,5 +21,15 @@ public ListTeamsResponse ExecuteGet(ListTeamsRequest request)
 
             return new ListTeamsResponse() { Teams = EntityFactory.ToDomain(teams) };
         }
+
+        public TeamResponse ExecutePost(CreateTeamRequest request)
+        {
+
+            // todo check if team name already exists, if so throw an error
```
Comment (Renny4Real) [inline SocialCareCaseViewerApi/V1/UseCase/TeamsUseCase.cs:28]: Would we not want to do this before releasing the ability to create a team? 
Just wondering if there's the risk of duplicate team names being created and then having to sift through them later?

**Review by Renny4Real — COMMENTED** [review]
Looks good! Just a few comments/questions? 🙂

File: SocialCareCaseViewerApi/V1/UseCase/TeamsUseCase.cs (hunk 2 above)
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi/V1/UseCase/TeamsUseCase.cs:28]: There shouldn't be any risk, GetTeams with either team id or team name is following this PR and they should both be in staging together at the same time.

We can wait to merge this until the other PR is ready to go though.

Also the frontend for this isn't done so it's fairly unlikely this backend API is hit by any users for now.

File: SocialCareCaseViewerApi/V1/Boundary/Requests/CreateTeamRequest.cs (hunk 1 above)
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi/V1/Boundary/Requests/CreateTeamRequest.cs:17]: Yes this is from FluentValidation. I believe there is a way to tell Startup.cs to automatically inject all our validators that implement AbstractValidator which would be nice. Something to look into to when I got some free time.

File: SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs [hunk 3]
```diff
@@ -385,6 +385,17 @@ public void CreateWorkerWithDuplicateEmailThrowsPostWorkerException()
             return (createWorkerRequest, createdTeams);
         }
 
+        [Test]
+        public void CreateTeamInsertsTeamIntoDatabaseAndReturnsCreatedWorker()
```
[the remark this answers is not in the document]
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi.Tests/V1/Gateways/DatabaseGateway.Tests.cs]: Nice catch

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/CreateTeamRequestTests.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        [Test]
+        public void ValidCreateTeamRequestReturnsNoErrorsOnValidation()
+        {
+            var createTeamRequest = TestHelpers.CreateTeamRequest();
+            var validator = new CreateTeamRequestValidator();
+
+            var validationResponse = validator.Validate(createTeamRequest);
+
+            if (validationResponse == null)
```
[the remark this answers is not in the document]
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi.Tests/V1/Boundary/Request/CreateTeamRequestTests.cs]: This wasn't needed!

Thanks and removed.

File: SocialCareCaseViewerApi.Tests/V1/Boundary/Request/CreateTeamRequestTests.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            {
+                (TestHelpers.CreateTeamRequest(name: ""), "Team name must be provided"),
+                (TestHelpers.CreateTeamRequest(name: _faker.Random.String2(201)), "Team name has a maximum length of 200 characters"),
+                (TestHelpers.CreateTeamRequest(context: ""), "Context flag must be provided\nContext flag must be 'A' or 'C'"),
+                (TestHelpers.CreateTeamRequest(context: "B"), "Context flag must be 'A' or 'C'"),
+            };
+
+            var validator = new CreateTeamRequestValidator();
+
+            foreach (var (request, expectedErrorMessage) in badCreateTeamRequests)
```
[the remark this answers is not in the document]
Comment (JohnFarrellDev) [reply SocialCareCaseViewerApi.Tests/V1/Boundary/Request/CreateTeamRequestTests.cs:34]: Your understanding is pretty much spot on.

Just a list of Requests created in a way that is invalid and the associated error message.

Test loops over the whole list checking that the validation fails for each bad request and the error message is what we expect.

**Review by Renny4Real — APPROVED** [review]
:shipit:

**Pull request merged**